src/serial.c

```diff
@@ -0,0 +1,236 @@
+#include <linux/kvm.h>
+#include <linux/serial_reg.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/select.h>
+#include <unistd.h>
+
+#include "serial.h"
+#include "vm.h"
+
+#define SERIAL_IRQ 4
+#define IO_READ8(data) *((uint8_t *) data)
+#define IO_WRITE8(data, value) ((uint8_t *) data)[0] = value
+
+#define container_of(ptr, type, member)               \
+    ({                                                \
+        void *__mptr = (void *) (ptr);                \
+        ((type *) (__mptr - offsetof(type, member))); \
+    })
+
+struct serial_dev_priv {
+    uint8_t dll;
+    uint8_t dlm;
+    uint8_t iir;
+    uint8_t ier;
+    uint8_t fcr;
+    uint8_t lcr;
+    uint8_t mcr;
+    uint8_t lsr;
+    uint8_t msr;
+    uint8_t scr;
+
+    uint8_t rx[FIFO_LEN];
```

Comment:
Ditto. We can create a straightforward FIFO implementation and place it in `src/util.h`.